### PR TITLE
Add recursive listing to roles

### PR DESCRIPTION
**Pull request by jefferai**
Notes:

* Nearly all of the logic in role_service.go can (and will) be refactored into a shared function when I sweep through the rest of the resource types
* I removed the additional verification logic because it's not used anywhere and the anticipated need for it has so far been satisfied by the later ActionSet functions. So I'm considering it dead code. It can always be revived from the repo history.
* There are some remaining tests to be written

### CODE REVIEW

**Comment by jefferai** [comment]
Taking out of draft so it can be commented on, but see notes above

File: internal/iam/repository_role.go [hunk 1]
```diff
@@ -182,13 +182,13 @@ func (r *Repository) DeleteRole(ctx context.Context, withPublicId string, _ ...O
...
+func (r *Repository) ListRoles(ctx context.Context, withScopeIds []string, opt ...Option) ([]*Role, error) {
 	const op = "iam.(Repository).ListRoles"
-	if withScopeId == "" {
-		return nil, errors.New(errors.InvalidParameter, op, "missing scope id")
+	if len(withScopeIds) == 0 {
+		return nil, errors.New(errors.InvalidParameter, op, "missing scope ids")
 	}
 	var roles []*Role
-	err := r.list(ctx, &roles, "scope_id = ?", []interface{}{withScopeId}, opt...)
+	err := r.list(ctx, &roles, "scope_id in (?)", []interface{}{withScopeIds}, opt...)
```
Comment (jimlambrt) [inline internal/iam/repository_role.go:191]: I'm interested to see how this bit o'magic works in gorm v2.  I have a suspicion that we may need to expand the "?" to match the number of paramters, but that's just a suspicion.. we'll see.  With that said, this is so deceptively simple.

**Review by jimlambrt — APPROVED** [review]
Looks great!  Lots of plumbing to do now.  Excellent work.  (PS. don't forget to add the needed tests at some point).

File: internal/iam/repository_role.go [hunk 2]
```diff
@@ -182,13 +182,13 @@ func (r *Repository) DeleteRole(ctx context.Context, withPublicId string, _ ...O
 }
 
 // ListRoles in a scope and supports WithLimit option.
-func (r *Repository) ListRoles(ctx context.Context, withScopeId string, opt ...Option) ([]*Role, error) {
+func (r *Repository) ListRoles(ctx context.Context, withScopeIds []string, opt ...Option) ([]*Role, error) {
```
Comment (mgaffney) [inline internal/iam/repository_role.go:185]: Can you update the godoc for the method to reflect the changes in the signature?

**Review by mgaffney — CHANGES_REQUESTED** [review]
Overall looks good. Just one minor comment about updating the godoc.

File: internal/iam/repository_role.go (hunk 2 above)
Comment (jefferai) [reply internal/iam/repository_role.go:185]: Sure, done.

**Comment by jefferai** [comment]
Added test and merged

**Pull request merged**